### ISSUE TITLE
--watch triggers --noUnusedLocals false positives

### ISSUE THREAD

**Issue by jwbay** [Microsoft/TypeScript#21518]
**TypeScript Version:**  2.7.1

**Search Terms:** 
watch mode, unusedlocals, declared but never read

**Code**

Run tsc --watch with this:
```ts
function one() {}

function two() {
  return function three() {
    one();
  }
}
```

Then change it to this:
```ts
function one() {}

// export added
export function two() {
  return function three() {
    one();
  }
}
```

**Expected behavior:**
No errors introduced

**Actual behavior:**
```
test.tsx(1,10): error TS6133: 'one' is declared but its value is never read.
17:32:24 - Compilation complete. Watching for file changes.
```

In a larger project, I saw this just from running tsc --watch, didn't need to make a change to trigger it.

This doesn't show up when running tsc without watch, and no errors show up in VSCode either.

**Comment by pelotom** [comment]
Here's my repro, which has an error on the first run with `tsc --watch`:

```ts
import _fromPairs from 'lodash-es/fromPairs'

export const fromPairs = <O>(pairs: [keyof O, O[keyof O]][]): O => (_fromPairs as any)(pairs)
```

**Comment by pelotom** [comment]
This is also affecting the latest nightly (`typescript@2.8.0-dev.20180131`).

**Comment by sheetalkamat** [comment]
@mhegazy this is because in watch mode to get declaration output (for verification if the signature is changed) we do not get diagnostics and later when we try to get diagnostics to report, the return expression result is already cached and hence we dont check it again but the expression doesnt get marked referenced. For now we should just enable the diagnostics again and investigate this better at later time so that declaration output doesn't impact diagnostics
